mdsuite/calculators/einstein_distinct_diffusion_coefficients.py

```diff
@@ -170,61 +202,24 @@ def _compute_msd(self, data: dict, data_path: list, combination: tuple):
+                Data paths for accessing the dictionary.
         Returns
         -------
         updates the class state
         """
-        for ensemble in tqdm(range(self.ensemble_loop), ncols=70, desc=str(combination)):
-            start = ensemble * self.correlation_time
-            stop = start + self.data_range
-            msd_a = self._msd_operation(
-                data[str.encode(data_path[0])][:, start:stop], square=False
-            )
-            msd_b = self._msd_operation(
-                data[str.encode(data_path[0])][:, start:stop], square=False
-            )
-
-            for i in range(len(data[str.encode(data_path[0])])):
-                for j in range(i + 1, len(data[str.encode(data_path[1])])):
-                    if i == j:
-                        continue
-                    else:
-                        self.msd_array += self.prefactor * np.array(
-                            tf.reduce_sum(msd_a[i] * msd_b[j], axis=1)
-                        )
-
-    def run_experimental_analysis(self):
-        """
-        Perform the distinct coefficient analysis analysis
-        """
-        if type(self.atom_selection) is dict:
-            select_atoms = {}
-            for item in self.atom_selection:
-                select_atoms[
-                    str.encode(join_path(item, "Unwrapped_Positions"))
-                ] = self.atom_selection[item]
-            self.atom_selection = select_atoms
-        for combination in self.combinations:
-            type_spec = {}
-            self._calculate_prefactor(combination)
-            data_path = [join_path(item, "Unwrapped_Positions") for item in combination]
-            self._prepare_managers(data_path=data_path)
-            type_spec = self._update_species_type_dict(type_spec, data_path, 3)
-            type_spec[str.encode("data_size")] = tf.TensorSpec(None, dtype=tf.int16)
-            batch_generator, batch_generator_args = self.data_manager.batch_generator(
-                dictionary=True
-            )
-            data_set = tf.data.Dataset.from_generator(
-                batch_generator, args=batch_generator_args, output_signature=type_spec
-            )
-            data_set = data_set.prefetch(tf.data.experimental.AUTOTUNE)
-            for batch in data_set:
-                self._compute_msd(batch, data_path, combination)
-            self._apply_averaging_factor()
-            self._post_operation_processes(combination)
+        msd_a = self.msd_operation(data[data_path[0]], square=False)
+        msd_b = self.msd_operation(data[data_path[0]], square=False)
+
+        for i in range(len(data[data_path[0]])):
+            for j in range(i + 1, len(data[data_path[1]])):
+                if combination[0] == combination[1] and i == j:
+                    continue
+                else:
+                    self.msd_array += self.prefactor * np.array(
+                        tf.reduce_sum(msd_a[i] * msd_b[j], axis=1)
+                    )
```

Comment:
It is certainly important for performance and definitely can be reduced to a summation over everything and then subtraction of diagonals but in the above einsum I am a little confused with where your tensor contractions are taking place. Could you provide an array example of what this looks like because by tensor contraction laws this should be a sum over x and y and the i and j should remain which is not quite what we want..

Reply:
In any case, there are two or three approaches I want to try for a speedup in the distinct coefficients. It may be better if I open an issue for improving correlation coefficient speed and when we have some nice test data play around with improving performance.

Reply:
Issue sounds good - we can collect potential methods there and  compare to existing data. Can be resolved when the issue is made.